INT-1759 Using same order reference ID when an error occurs

rlopz opened the pull request:
[INT-1759](https://jira.bigcommerce.com/browse/INT-1759)
## What?
We started retrieving the orderReferenceId from order to use the same value when InvalidPaymentMethod or MFA-Abandoned occurs we do not want to generate another orderReferenceId in this cases

## Why?
We do not want to hit the amazon endpoint for generating another orderReferenceId because we already have one

## Siblings PR
[INT-1744 BCAPP](https://github.com/bigcommerce/bigcommerce/pull/31095)

## Testing / Proof
![Screen Shot 2019-08-26 at 12 45 39 PM](https://user-images.githubusercontent.com/39630835/63710754-88f73100-c7ff-11e9-98d5-bb00a7747268.png)
![Screen Shot 2019-08-26 at 12 47 01 PM](https://user-images.githubusercontent.com/39630835/63710797-a2987880-c7ff-11e9-9a32-5cf575ab8283.png)




@bigcommerce/checkout @bigcommerce/payments @bigcommerce/intersys-integrations

rlopz commented:
unit testing will be added after this approach is approved

metalvegetarianoprogresivo commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.ts:
```diff
@@ -138,10 +138,17 @@ export default class AmazonPayPaymentStrategy implements PaymentStrategy {
         return new Promise((resolve, reject) => {
             const { container, onError = noop, onPaymentSelect = noop, onReady = noop } = options;
-            const referenceId = this._getOrderReferenceId();
+            const referenceId = this._getOrderReferenceId() || this._getOrderReferenceIdFromOrder();
```
What happens if both ar `undefined`?

rlopz replied:
if `this._getOrderReferenceId()` comes undefined another `orderReferenceId` will be generated

metalvegetarianoprogresivo replied:
Okay, sound good.

metalvegetarianoprogresivo commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.ts:
```diff
@@ -256,7 +263,8 @@ export default class AmazonPayPaymentStrategy implements PaymentStrategy {
 
-                            return new Promise<never>(() => {});
+                            return new Promise<never>(() => {
+                            });
```
There's no point on splitting this in two lines.

metalvegetarianoprogresivo commented on src/order/order.ts:
```diff
@@ -29,6 +29,7 @@ export default interface Order {
     taxTotal: number;
     payments?: OrderPayments;
     status: string;
+    providerTransactionId: string;
```
Can you sort this alphabetically?

metalvegetarianoprogresivo commented on src/order/orders.mock.ts:
```diff
@@ -25,6 +25,7 @@ export function getOrder(): Order {
         hasDigitalItems: false,
         isComplete: true,
         status: 'ORDER_STATUS_AWAITING_FULFILLMENT',
+        providerTransactionId: 'providerTransactionId',
```
Can you sort this alphabetically?

rlopz commented:
@metalvegetarianoprogresivo I have addressed your comments, could you please take another look? ♻️

rlopz commented:
> Hi @rlopz, thanks for your PR, can you write a test to verify your change?
Hi @davidchin, the unit test that I made on the last PR related with 3ds on amazon is already covering this new code. 
<img width="792" alt="Screen Shot 2019-08-19 at 11 23 41 AM" src="https://user-images.githubusercontent.com/39630835/63283614-584f4e80-c277-11e9-8026-e7aaf7581457.png">
<img width="560" alt="Screen Shot 2019-08-19 at 11 23 27 AM" src="https://user-images.githubusercontent.com/39630835/63283620-5ab1a880-c277-11e9-8459-2b42810359c4.png">

rlopz commented:
Hi @davidchin, I just change some code in order to follow the suggestion from Sydney team, could you please take another look ♻️

davidchin commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.ts:
```diff
@@ -92,6 +94,10 @@ export default class AmazonPayPaymentStrategy implements PaymentStrategy {
         }
 
+        if (!this._isPaymentMethodSelected) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
```
Are you sure this is the right error to throw? As far I know, `_isPaymentMethodSelected` would be false if the shopper hasn't selected a credit card option, or we are not able to save the billing address provided by Amazon, not purely because the developer using this library has forgotten to `initialize` the payment method. You can either create a new custom error type that's more specific or use an existing one such as `PaymentMethodFailedError`.

rlopz replied:
Hi @davidchin thanks for the observation, I did not found the `PaymentMethodFailedError` but one that I think works perfectly for this case which is: `MissingPaymentMethod`

davidchin commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.ts:
```diff
@@ -167,7 +178,10 @@ export default class AmazonPayPaymentStrategy implements PaymentStrategy {
                     this._synchronizeBillingAddress()
-                        .then(() => onPaymentSelect(orderReference))
+                        .then(() => {
+                            this._isPaymentMethodSelected = true;
```
I believe `onPaymentSelect` gets called each time the shopper selects a new credit card option. Does that mean once it's set to `true`, there's a chance that something could fail later and the flag should be unset?

rlopz replied:
Yes you are right `onPaymentSelect` is called each time the shopper selects a new credit card option. Actually this flag was added in order to be able to identify when a credit card is not selected. This scenario occurs when the wallet widget is loaded after a purchase failed with an `InvalidPaymentMethod` code

davidchin commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.ts:
```diff
@@ -138,10 +144,15 @@ export default class AmazonPayPaymentStrategy implements PaymentStrategy {
     }
 
+    private _getOrderReferenceIdFromInitializationData(): string | undefined {
+
```
Unnecessary line break. 🍹

davidchin commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.spec.ts:
```diff
@@ -470,26 +471,10 @@ describe('AmazonPayPaymentStrategy', () => {
     });
 
     describe('When 3ds is enabled', () => {
```
I still can't see a test that verifies you're using the order reference from the `initializationData` as a fallback.

rlopz commented:
Hi @davidchin, I have addressed your comments ♻️, could you please take another look? also we have detected that exceptions are failing on execute methods not only for `Amazon` even we tested with `Authorizenet` and same error is happening, for example when `PaymentNotInitialized` is threw a spinner remains loading

rlopz commented on src/payment/strategies/amazon-pay/amazon-pay-payment-strategy.ts:
```diff
@@ -85,11 +87,15 @@ export default class AmazonPayPaymentStrategy implements PaymentStrategy {
 
         if (!referenceId || !sellerId) {
-            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+            return Promise.reject(new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized));
```
Done! Thank you

rlopz commented:
Hi @davidchin I addressed your comments. Could you review it again?

PascalZajac commented:
@rlopz the commit message and PR title mention different things, can you please amend them appropriately? Then we'll merge this. Please ensure they reflect _what_ changed, not ticket titles.

rlopz commented:
> @rlopz the commit message and PR title mention different things, can you please amend them appropriately? Then we'll merge this. Please ensure they reflect _what_ changed, not ticket titles.

Hi @PascalZajac I just have updated both texts, could you please take another look? ♻️

PascalZajac commented:
@rlopz one last request, can you change the scope in the commit message from `checkout` to `payment`?
